src/hotspot/share/runtime/handshake.cpp

```diff
@@ -497,8 +497,17 @@ HandshakeOperation* HandshakeState::get_op_for_self(bool allow_suspend, bool che
 }
 
 bool HandshakeState::has_operation(bool allow_suspend, bool check_async_exception) {
-  MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
-  return get_op_for_self(allow_suspend, check_async_exception) != nullptr;
+  // We musn't block here as that could lead to deadlocks if we already hold an
+  // "external" mutex. If the try_lock fails then we assume that there is an operation
+  // and force the caller to check more carefully in a safer context. If we can't get
+  // the lock it means another thread is trying to handshake with us, so it can't
+  // happen during thread termination and destruction.
```

Comment:
Why the particular mention about thread termination and destruction?

Reply:
nit typo: s/musn't/mustn't/

which is preferred according to urbandictionary.com.

I also don't understand the last sentence. More accurately, I understand the first
phrase, but I don't understand the need for the second phrase.

Reply:
Earlier when I mistakenly thought the no-arg version was this version with default args, I was concerned about its use in the `HandshakeState` destructor, but reasoned that it was safe as there couldn't be a real pending op at that time to cause contention on the lock - hence the comment in the main code. I left the comment just to add some information on when during a thread's lifetime we could hit this problem - to make it easier to reason about. I can remove it if it is causing confusion.

Reply:
I see. The only thing which might be confusing is what exactly "termination" phase includes. Because when we are in Threads::remove() with a status of _thread_exiting we can still fail to acquire the lock here while trying to grab the Threads_lock.